keps/sig-node/4680-add-resource-health-to-pod-status/README.md

```diff
@@ -0,0 +1,408 @@
+	// enacted on the running container after it has been started or has been successfully resized.
+	// +featureGate=InPlacePodVerticalScaling
+	// +optional
+	Resources *ResourceRequirements `json:"resources,omitempty" protobuf:"bytes,11,opt,name=resources"`
+```
+
+One field reflects the resource requests and limits and the other actual allocated resources.
+
+This structure will contain standard resources as well as extended resources. As noted in the comment: https://github.com/kubernetes/kubernetes/pull/124227#issuecomment-2130503713, it is only logical to also include the status of those allocated resources. 
+
+The proposal is to keep this structure as-is to simplify parsing of well-known ResourceList data type by various consumers. Typical scenario would be to compare if the AllocatedResources match the desired state.
+
+The proposal is to introduce an additional field:
+
+```
+// AllocatedResourcesStatus represents the status of various resources
+// allocated for this Pod.
+AllocatedResourcesStatus ResourcesStatus 
+
+type ResourcesStatus map[ResourceName]ResourceStatus
+
+type ResourceStatus struct {
+    // map of unique Resource ID to its status
+    Resources map[ResourceID] ResourceStatus
+    
+    // allow to extend this struct in future with the overall health fields or things like Device Plugin version
+}
+
+type ResourceID string
+
+type ResourceStatus struct {
+    // can be one of:
+    //  - Healthy: operates as normal
+    //  - Unhealthy: reported unhealthy. We consider this a temporary health issue since we do not mechanism today to distinguish temporary and permanent issues.
+    //  - Unknown: The status cannot be determined. For example, Device Plugin got unregistered and hasn't been re-registered since.
+    Status string
+}
+```
+
+***Is there any guarantee that the AllocatedResourcesStatus will be updated before Container crashed and unscheduled?***
+
+No, there is no guarantee that the Device Plugin/DRA will detect device going unhealthy earlier than the Pod. Once device got unhealthy, container may crash and being marked as Failed already.
+
+The proposal is to update the Pod Status with the device status even if the pod has been marked as Failed already, but still known to the kubelet.
+
+
+***Do we need the CheckDeviceHealth call introduced to the Device Plugin to work around the limitation above?***
+
+We may consider this as a future improvement. 
+
+
+***Should we introduce a permanent failure status?***
+
+We may consider this as a future improvement. 
+
+### User Stories (Optional)
+
+#### Story 1
+
+- User scheduled a Pod using the GPU device
+- When GPU device fails, user sees the Pod is in crash loop backoff
+- User checks the Pod Status using `kubectl describe pod`
+- User sees the pod status indicating that the GPU device is not healthy
+- User or some (custom for now) controller deletes the Pod and replicaset reschedules it on another available GPU
```

Comment:
> (custom for now) controller deletes the Pod

If kubelet knows the device is unhealthy would it make sense to mark the pod as Failed? If the kubelet knows the device is unhealthy; what is the benefit of keeping the pod in a crash loop?

Reply:
I want to minimize changes taken in a single KEP. My reasoning here - what if device only needed during pod initialization (some crypto thing) and pod can run normally after? What if one device going down is not disrupting the whole Pod and it can keep running with limited capacity on smaller number of devices?

Changing behavior in this KEP will need to account for cases like above and I beliebe it will lead to larger discussion